nilearn/decoding/tests/test_decoder.py

```diff
@@ -39,60 +50,140 @@
+        n_samples=n_samples,
+        n_features=125,
+        scale=3.0,
+        n_informative=5,
+        n_classes=4,
+        random_state=42,
+    )
+    X, mask = to_niimgs(X, [5, 5, 5])
+    return X, y, mask
+
+
+@pytest.fixture(scope="session")
+def dummy_binary_classification_data():
+    """Use for testting errors."""
+    return _make_binary_classification_test_data(n_samples=20)
+
+
+@pytest.fixture
+def binary_classification_data():
+    """Use for testting errors."""
+    return _make_binary_classification_test_data(n_samples=N_SAMPLES)
+
+
+def _make_regression_test_data(n_samples=N_SAMPLES, dim=30):
+    X, y = make_regression(
+        n_samples=n_samples,
+        n_features=dim**3,
+        n_informative=dim,
+        noise=1.5,
+        bias=1.0,
+        random_state=42,
+    )
+    X = StandardScaler().fit_transform(X)
+    X, mask = to_niimgs(X, [dim, dim, dim])
+    return X, y, mask
+
+
+@pytest.fixture
+def regression_data():
+    return _make_regression_test_data(n_samples=N_SAMPLES, dim=30)
+
+
+@pytest.fixture
+def multiclass_data():
+    return _make_multiclass_classification_test_data(n_samples=N_SAMPLES)
+
+
+@pytest.mark.parametrize(
+    "regressor, param", [(RidgeCV(), []), (SVR(kernel="linear"), "C")]
+)
+def test_check_param_grid_regression(regressor, param):
     """Test several estimators.
 
     Each one with its specific regularization parameter.
     """
-    # Regression
-    for _, (regressor, param) in regressors.items():
-        param_grid = _check_param_grid(regressor, X, y_regression, None)
-        assert list(param_grid.keys()) == list(param)
-    # Classification
-    for _, (classifier, param) in classifiers.items():
-        param_grid = _check_param_grid(classifier, X, y_classification, None)
-        assert list(param_grid.keys()) == list(param)
-
-    # Using a non-linear estimator to raise the error
-    for estimator in ["log_l1", random_forest]:
-        pytest.raises(
-            ValueError, _check_param_grid, estimator, X, y_classification, None
-        )
+    rng = np.random.RandomState(0)
+    X = rng.rand(N_SAMPLES, 10)
+    Y = rng.rand(N_SAMPLES)
+
+    param_grid = _check_param_grid(regressor, X, Y, None)
+
+    assert list(param_grid.keys()) == list(param)
+
+
+@pytest.mark.parametrize(
+    "classifier, param",
+    [
+        (LogisticRegression(penalty="l1"), "C"),
+        (LogisticRegression(penalty="l2"), "C"),
+        (RidgeClassifierCV(), []),
+    ],
+)
+def test_check_param_grid_classification(rand_X_Y, classifier, param):
+    """Test several estimators.
+
+    Each one with its specific regularization parameter.
+    """
+    X, Y = rand_X_Y
+
+    param_grid = _check_param_grid(classifier, X, Y, None)
+
+    assert list(param_grid.keys()) == list(param)
+
+
+@pytest.mark.parametrize("estimator", ["log_l1", RandomForestClassifier()])
+def test_check_param_grid_error(rand_X_Y, estimator):
+    """Raise the error when using a non-linear estimator."""
```

Comment:
I think that the point is that this estimator does not take a param_grid ?

Reply:
I added a match in the pytest.raise to check the error message content.
Modified the test name as it actually only makes sure that unsupported estimators are rejected.